### DISCUSSION TITLE
Fl_Text_Editor scrolling while not in focus

### DISCUSSION THREAD

**Discussion by MoAlyousef** [fltk/fltk#1010] [Q&A] [answered]
**Describe the issue**
This discussion is formatted as if it were a bug since I'm not sure if it is or not. If it is, it can be moved into an issue.
This was reported in the fltk-rs repo here:
https://github.com/fltk-rs/fltk-rs/issues/1568

I was able to reproduce the same behavior in C++ as well.
Basically an Fl_Text_Editor is scrolled even when not having focus. When having for example 2 editors in an application, when one of the editors is scrolled to the end, the other editor will also start scrolling down. Similarly when scrolling to the top, when the scrollbar reaches the top, the other editor will also start scrolling.

**To Reproduce**
Build the following program:
```c++
#include <FL/Fl_Text_Editor.H>
#include <FL/Fl_Text_Buffer.H>
#include <FL/Fl_Double_Window.H>
#include <FL/Fl_Flex.H>
#include <FL/Fl.H>

int main(int argc, char *argv[]) {
  auto *win = new Fl_Double_Window(600, 400);
  auto *row = new Fl_Flex(0, 0, 600, 400);
  row->type(Fl_Flex::ROW);
  auto *ed1 = new Fl_Text_Editor(0, 0, 0, 0);
  auto *ed2 = new Fl_Text_Editor(0, 0, 0, 0);
  row->end();
  win->end();
  win->resizable(row);
  win->show(argc, argv);

  auto *buf1 = new Fl_Text_Buffer;
  auto *buf2 = new Fl_Text_Buffer;

  // fill buffers
  for (auto i = 0; i < 50; i++) buf1->append("Hello\n");
  for (auto i = 0; i < 50; i++) buf2->append("World\n");
  
  ed1->buffer(buf1);
  ed2->buffer(buf2);

  return Fl::run();
}
```

After launching the program, it displays a window with 2 editors, the first editor on the left will have focus, start scrolling downwards using the mousewheel, observe that the right editor will start scrolling when the left editor stops scrolling.

**Expected behavior**
The right editor shouldn't start scrolling since it has no focus.

**Screenshots**
N/A

**FLTK Version**
 - Version: 1.4.0
 - If from Git, branch: master

**FLTK Configure / Build Options**
```bash
cmake -Bbin -DFLTK_BUILD_TEST=OFF && cmake --build bin
```

**Operating System / Platform:**
Please be as precise as possible, e.g. "Linux: Ubuntu 20.04"
 - OS: Linux on WSL. Running Ubuntu 22.04.3 LTS. Windows 11 x86_64.
 - The linked issue also observes the same behavior on macos.

**Linux/Unix Runtime, if applicable:**
 - X11

**Comment by Albrecht-S** [comment, marked as answer]
@MoAlyousef I believe this behavior is normal for FLTK applications and also for other applications.

1. Scroll wheel events are not delivered (primarily) to the **_focus widget_** but to the **_widget below the mouse_**. In the given demo program you can see this by moving the mouse over one or the other editor while both scrollbars are in a "middle" position and then scrolling one or the other editor.
2. When the editor that's being scrolled reaches the end in the current direction it does no longer accept the scroll event and this event is propagated to its parent - in the example program the `Fl_Flex` widget.
3. This widget propagates the event to its children until one of them (in this case the "other" `Fl_Text_Editor` widget) accepts it, i.e. returns 1, hence the other editor widget scrolls.
4. You can see this in other applications as well, for instance in GitHub's Issue or Discussions web interface: fill some lines of text into the "answer" box (more than can be displayed) so the scrollbar appears. With the cursor at the end of the text start scrolling upwards. As soon as the text in the input box reaches the top the "parent", i.e. the content of the "discussion" page (question and previous answers) begins to scroll up.
5. In the same example (4.) you can also see that the scroll event is delivered to the "widget" below the mouse, not the widget that has focus: leave the focus in the text input box, move the mouse outside of it, and start scrolling: notice how the page contents scrolls although the text input box (the "editor widget") has the focus.

That said, the described behavior is how FLTK event delivery works. The only point that may be considered questionable in this particular case would be whether the parent (`Fl_Flex`) widget should propagate the event to its children and thus scroll the "other" editor widget that has neither the focus nor is below the mouse. However, this would be hard to answer, even if we wanted to do it (better?).

Thus I consider this behavior as intended although it **_might_** be **in parts** disputable.

PS: AFAICT this entire behavior is platform independent. I could confirm it on Linux/Wayland as well (Windows not tested).

**Reply by MoAlyousef** [reply]
Thanks @Albrecht-S for looking into this.
Yeah I guess scrolling when the mouse is above the widget would be acceptable. I agree with you that its questionable that the second editor also scrolls when not in focus and not under the mouse.

**Reply by Albrecht-S** [reply]
@MoAlyousef Although I wrote that it's questionable (or disputable) I don't think that there's any chance that we could change it. This is how FLTK's event delivery works and it's baked into the core of FLTK. I won't attempt to change this, at least not in 1.4.0, but maybe other devs have an idea (@MatthiasWM ?).

That said, in FLTK 1.5.0 we will get touch events (zoom, scroll, etc.) that have similar issues (the location of the event needs to be considered). It's hard to tell how touch event delivery is supposed to work (I tried it already) and then to redesign or modify global event delivery to meet these requirements. Maybe this will also affect mousewheel events and we'll find a way to do it properly in 1.5.0.

**Reply by MoAlyousef** [reply]
Thanks for clarifying further @Albrecht-S.
